google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ServerToIDEFileResolver.java

```diff
@@ -0,0 +1,124 @@
+      new ConcurrentHashMap<Project, Map<String, VirtualFile>>();
+  private VirtualFileSystem fileSystem =
+      VirtualFileManager.getInstance().getFileSystem(StandardFileSystems.FILE_PROTOCOL);
+
+  public ServerToIDEFileResolver(@NotNull final Project project, boolean forceRefresh) {
+    initializeLocationsForProject(project, forceRefresh);
+  }
+
+  /**
+   * Utility method that returns the full class name for a file.
+   */
+  public static String getCloudPathFromJavaFile(PsiJavaFile javaFile) {
+    return javaFile.getPackageName().replace('.', '/') + "/" + javaFile.getName();
+  }
+
+  /**
+   * Given a possible file path, returns a VirtualFile instance that the IDE can work with.
+   *
+   * <p>This method tries to fetch a file in three ways. First, it uses its full path in the local
+   * file system. Then, it tries the full class name form (com/google/gct/idea/debugger/
+   * CloudDebugProcess.java). Finally, it searches for possible file matches within the project.
+   */
+  public VirtualFile getFileFromPath(@NotNull Project project, @NotNull String path) {
+    // Try the relative full project path.
+    VirtualFile file = fileSystem.findFileByPath(project.getBasePath() + "/" + path);
+    // Try class name with package and class file name.
+    if (file == null) {
+      file = ourLocationMaps.get(project).get(path);
+    }
+    // If we still couldn't find the file, search for possible file name matches and return the
+    // first.
+    // We might want to improve string matching and return more than one possible match.
+    if (file == null) {
+      Collection<VirtualFile> projectJavaFiles = FileBasedIndex.getInstance().getContainingFiles(
+          FileTypeIndex.NAME, JavaFileType.INSTANCE, GlobalSearchScope.projectScope(project));
+      for (VirtualFile projectFile : projectJavaFiles) {
+        if (projectFile.getName().equals(path)) {
+          file = projectFile;
+          break;
+        }
+      }
+    }
+    return file;
+  }
+
+  /**
+   * Creates a map from package-path syntax to {@link VirtualFile} which allows us to build IDE
+   * source locations given information from the Cloud Debugger API. We only do a force refresh when
+   * re-attaching to a target because the project state may change between debug sessions.
+   */
+  public void initializeLocationsForProject(@NotNull final Project project, boolean forceRefresh) {
+    if (forceRefresh || !ourLocationMaps.containsKey(project)) {
+      final Map<String, VirtualFile> locationMap =
+          Maps.newHashMap(); // The inner does not have to be concurrent.
+      Collection<VirtualFile> projectJavaFiles = FileBasedIndex.getInstance().getContainingFiles(
+          FileTypeIndex.NAME, JavaFileType.INSTANCE, GlobalSearchScope.projectScope(project));
+      for (VirtualFile virtualFile : projectJavaFiles) {
+        PsiFile javaFile = PsiManager.getInstance(project).findFile(virtualFile);
```

Comment:
Couldn't you centralize the transform in one method and call that from everywhere you check the map?